sfge/src/main/java/com/salesforce/graph/ops/expander/ApexPathCollapserImpl.java

```diff
@@ -16,32 +16,51 @@
 final class ApexPathCollapserImpl implements ApexPathCollapser {
     private static final Logger LOGGER = LogManager.getLogger(ApexPathCollapserImpl.class);
 
+    private final PathExpansionRegistry registry;
+    private final Long pathExpansionId;
+
     /** This is the list of collapsers that the user has requested */
     private final List<ApexDynamicPathCollapser> dynamicPathCollapsers;
 
     /**
-     * Maps a ForkEvent to all of the ApexPathExpanders that have the same ForkEvent and returned a
-     * result. The ForkEvent passed to {@link #resultReturned(ApexPathExpander, ForkEvent,
-     * Optional)} are used to find all other ApexPathExpanders that might potentially be collapsed.
+     * Maps a ForkEvent Id to all of the ApexPathExpander Ids that have the same ForkEvent and
+     * returned a result. The ForkEvent passed to {@link #resultReturned(ApexPathExpander,
+     * ForkEvent, Optional)} are used to find all other ApexPathExpanders that might potentially be
+     * collapsed.
      */
-    private final Map<ForkEvent, List<ApexPathExpander>> forkEventToApexExpandersWithResults;
+    private final Map<Long, List<Long>> forkEventIdToApexExpanderIdsWithResults;
 
-    /** The list of all expanders that ApexPathExpanderUtil should remove on its next iteration */
-    private final List<ApexPathExpander> collapsedExpanders;
+    /**
+     * The list of Ids of all expanders that ApexPathExpanderUtil should remove on its next
+     * iteration
+     */
+    private final List<Long> collapsedApexPathExpanderIds;
 
-    ApexPathCollapserImpl(List<ApexDynamicPathCollapser> dynamicPathCollapsers) {
+    ApexPathCollapserImpl(
+            Long pathExpansionId,
+            List<ApexDynamicPathCollapser> dynamicPathCollapsers,
+            PathExpansionRegistry registry) {
+        this.pathExpansionId = pathExpansionId;
         this.dynamicPathCollapsers = dynamicPathCollapsers;
-        this.forkEventToApexExpandersWithResults = new HashMap<>();
-        this.collapsedExpanders = new ArrayList<>();
+        this.registry = registry;
+        this.forkEventIdToApexExpanderIdsWithResults = new HashMap<>();
+        this.collapsedApexPathExpanderIds = new ArrayList<>();
         if (dynamicPathCollapsers.isEmpty()) {
             throw new UnexpectedException("Use NoOpApexPathCollapser");
         }
     }
 
+    @Override
+    public Long getId() {
+        return pathExpansionId;
+    }
```

Comment:
Every ApexPath path expansion creates only a single instance of ApexPathCollapser. However, previously, ApexPathExpander held onto an instance of the ApexPathCollapser, and ApexPathCollapser maintained large collections of ApexPathExpander. By adding ApexPathCollapser to the registry, I'm trying to reduce the tight coupling. 

However, I'm now reconsidering if we'll still have all the optimization benefits even if this was not done. If I have the time, I'll try to undo this part and test its effect on heap usage.